AdyenCard/Utilities/BinLookup/BinLookupService.swift

```diff
@@ -0,0 +1,59 @@
+//
+// Copyright (c) 2020 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import Foundation
+
+/// Provide cardType detection based on BinLookup API.
+internal protocol AnyBinLookupService {
+    func requestCardType(for bin: String, caller: @escaping (BinLookupResponse) -> Void)
+}
+
+/// Provide cardType detection based on BinLookup API.
+/// Fall back to local regex-based detector if API not available or BIN too short.
+internal final class BinLookupService: AnyBinLookupService {
+    
+    private static let MinBinLength = 7
+    
+    private lazy var cardTypeDetector: CardTypeDetector = CardTypeDetector(detectableTypes: self.supportedCardTypes)
+    private let supportedCardTypes: [CardType]
+    private let publicKey: String
+    private let clientKey: String
+    private let apiClient: APIClientProtocol
+    
+    internal init(supportedCardTypes: [CardType], environment: APIEnvironment, publicKey: String, clientKey: String) {
+        self.supportedCardTypes = supportedCardTypes
+        self.publicKey = publicKey
+        self.clientKey = clientKey
+        self.apiClient = APIClient(environment: environment)
+    }
+    
+    /// Request card types based on enterd BIN.
+    /// - Parameters:
+    ///   - bin: Card's BIN number. If longer than `MinBinLength` - calls API, otherwise check local Regex,
+    ///   - caller:  Callback to notify about results.
+    internal func requestCardType(for bin: String, caller: @escaping (BinLookupResponse) -> Void) {
```

Comment:
can we use `Result<[CardType], Error>` instead?, the fact that `BinLookupService` does a network request is private logic IMO.

Reply:
Yep, I thought about it.. there is more fields coming from `BinLookupResponse` and it possibly could grow in time.

Probably, I can split it into CardTypeProivider(for hight level card type detection) and BINLookupService (for throttling and network calls).